File: cpp-scheduling-algorithms/src/shortest_job_first.cpp
#include "process.hpp"
#include <vector>
#include <queue>
#include <algorithm>
#include <unordered_map>

class ShortestJobFirst {

    std::vector<Process> processes;

    public:


    ShortestJobFirst( std::vector<Process> processes ) {
        this -> processes = processes;
    }


    void schedule() {

        // For preemptive scheduling algorithms, the burstTime of `processes`
        // will be modified in further steps,
        // hence we store the user-given burstTimes in a hashmap
        // for later use
        std::unordered_map<std::string,long> burstTimes;
        for( Process p : processes ) {
            burstTimes[ p.name ] = p.burstTime ; 
        }

        std::vector<Process> readyQueue; // Use vector as a queue here
        std::vector<Process> output;    // Stores completed processes

        long time = 0 ;
        bool isExecuting = false ;      // A flag to check if a process is executing currently
        Process currentProcess = processes[0];  // Stores the currently executing process
        int n = 0;      // Indicates number of completed processes. Equals the size of `output`
        int step = 0 ;    // Indicates the time when the currentProcess should be preempted
                          // and a new process should be scheduled
        long currentProcessRemainingBT = 1000L;     // The burstTime (execution time) remaining
        // for currentProcess. It is decremented by 1 at each time-step.

        while( n < processes.size() ) {

            // Add processes to ready queue if any arrives
            // To check if a process has arrived, match
            // `time` with each process's arrivalTime
            for( Process p : processes ) {
                if( p.arrivalTime == time ) {
                    readyQueue.push_back( p ) ; 
                }
            }

            // `currentProcessRemainingBT` becomes 0 when the `currentProcess`
            // has finished its execution. In this case, we append `currentProcess` to
            // `output` after calculating the required parameters
            if( currentProcessRemainingBT == 0L ) {

                // Insert `currentProcess` in `output`
                // and set isExecuting=false as the CPU becomes idle
                currentProcess.responseTime -= currentProcess.arrivalTime;
                currentProcess.burstTime = burstTimes[currentProcess.name];
                currentProcess.completionTime = time ; 
                currentProcess.turnAroundTime = currentProcess.completionTime - currentProcess.arrivalTime;
                currentProcess.waitTime = currentProcess.turnAroundTime - currentProcess.burstTime;
                output.push_back( currentProcess ) ; 
                n++ ;
                isExecuting = false ; 

                // If there exists a process in the readyQueue,
                // schedule it for execution as the CPU is idle
                if( !readyQueue.empty() ) {
                    // Sort processes in the readyQueue according to
                    // their burstTimes in increasing order.
                    std::sort( readyQueue.begin() , readyQueue.end() , []( const Process& p1 , const Process& p2 ){
                        return p1.burstTime < p2.burstTime ;
                    });
                    // The first process in the readyQueue will have the least burstTime
                    // Set it as the currentProcess and change `isExecuting` to true
                    if( readyQueue[0].responseTime == 0L ) readyQueue[0].responseTime = time;
                    currentProcessRemainingBT = readyQueue[0].burstTime ; 
                    currentProcess = readyQueue[0];
                    readyQueue.erase( readyQueue.begin() ) ; 
                    isExecuting = true;
                }
            }
            else {
                if( !readyQueue.empty() ) {
                    // If the readyQueue is not empty,
                    // sort processes in the readyQueue according to their burstTimes
                    std::sort( readyQueue.begin() , readyQueue.end() , []( const Process& p1 , const Process& p2 ){
                        return p1.burstTime < p2.burstTime ;
                    });
                    if( isExecuting ) {
                        // readyQueue[0] is the process with least burstTime in the readyQueue
                        // If the burstTime of that process is smaller
                        // than the remaining burstTime of the currently executing process,
                        // then preempt the currentProcess
                        // and start execution of readyQueue[0]
                        if( readyQueue[0].burstTime < currentProcessRemainingBT ) {
                            currentProcess.burstTime = currentProcessRemainingBT ; 
                            readyQueue.push_back( currentProcess ) ; 
                            currentProcess = readyQueue[ 0 ] ; 
                            readyQueue.erase( readyQueue.begin() ) ; 
                            currentProcessRemainingBT = currentProcess.burstTime ; 
                            isExecuting = true;
                        } 
                    }
                    else {
                        // If no process was executing (hence no process to preempt)
                        // schedule process readyQueue[0]
                        currentProcess = readyQueue[ 0 ] ; 
                        readyQueue.erase( readyQueue.begin() ) ; 
                        currentProcessRemainingBT = currentProcess.burstTime ; 
                        isExecuting = true;
                    }
                
                }
            }

            // Decrement currentProcessRemainingBT only if a process
            // is executing currently
            if( isExecuting ) {
                currentProcessRemainingBT-- ; 
            }

            time++ ; 
        }

        // Compute average turnAroundTime and waitTime
        float avgTAT = 0.0f ; 
        float avgWT = 0.0f;
        for( Process p : output ) {
            avgTAT += p.turnAroundTime ; 
            avgWT += p.waitTime;
            std::cout << p << "\n" ; 
        }
        avgTAT /= output.size() ; 
        avgWT /= output.size() ; 

        std::cout << "Average TAT: " << avgTAT << "\n" ; 
        std::cout << "Average WT: " << avgWT << "\n" ;  
    }

} ;

File: py-scheduling-algorithms/shortest_job_first.py
from collections import deque
from process import Process

class ShortestJobFirst:

    @classmethod
    def schedule( cls, processes: list[Process] ):

        burst_times: dict[int,int] = {}
        for proc in processes:
            burst_times[ proc.id ] = proc.bt

        ready_queue = deque()
        schedule: list[Process] = []

        time: int = 0 
        n_completed_proc: int = 0
        curr_proc: Process = None
        curr_proc_bt: int = 1e+4
        is_proc_exec: bool = False 

        while n_completed_proc < len( processes ):

            # 1. Add processes to ready_queue if any has arrived
            for proc in processes:
                if proc.at == time:
                    ready_queue.append( proc )

            # 2. Check for completion of current process
            if is_proc_exec and curr_proc_bt == 0:
                curr_proc.ct = time
                curr_proc.bt = burst_times[ curr_proc.id ]
                curr_proc.calc_params()
                schedule.append( curr_proc )
                n_completed_proc += 1
                is_proc_exec = False

            # 3. Check if current process can be preempted
            if is_proc_exec and len( ready_queue ) > 0:
                min_bt_proc = min( ready_queue , key=lambda proc: proc.bt ) 
                if min_bt_proc.bt < curr_proc_bt:
                    # Preempt the current process
                    curr_proc.bt = curr_proc_bt 
                    ready_queue.append( curr_proc )
                    curr_proc = min_bt_proc
                    curr_proc_bt = curr_proc.bt
                    ready_queue.remove( min_bt_proc )

            # 4. Schedule a process from ready_queue if CPU is idle
            if not is_proc_exec and len( ready_queue ) > 0:
                min_bt_proc = min( ready_queue , key=lambda proc: proc.bt ) 
                ready_queue.remove( min_bt_proc )
                curr_proc = min_bt_proc
                curr_proc.rt = time
                curr_proc_bt = curr_proc.bt
                is_proc_exec = True

            if is_proc_exec:
                curr_proc_bt -= 1

            time += 1

        avg_tat = sum( [ proc.tat for proc in schedule ] ) / len( processes )
        avg_wt = sum( [ proc.wt for proc in schedule ] ) / len( processes )
        print( f"AVG TAT: {avg_tat}" )
        print( f"AVG WT: {avg_wt}" )
        for proc in schedule:
            print( proc )

bts : list[int] = [ 6 , 2 , 8 , 3 , 4  ] 
ats : list[int] = [ 2 , 5 , 1 , 0 , 4  ]
processes: list[Process] = []
for i , ( at , bt ) in enumerate( zip( ats , bts ) ):
    proc = Process()
    proc.id = i + 1
    proc.at = at
    proc.bt = bt
    processes.append( proc )

ShortestJobFirst.schedule( processes )


